src/Cardano/Metadata/Validation/Wallet.hs

```diff
@@ -0,0 +1,91 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+{- | Wallet-specific validation rules.
+-}
+
+module Cardano.Metadata.Validation.Wallet
+  ( WalletValidationError(..)
+  , prettyPrintWalletValidationError
+  , walletValidationRules
+  , walletValidation
+  ) where
+
+import Cardano.Metadata.GoguenRegistry
+    ( PartialGoguenRegistryEntry, parseRegistryEntry, validateEntry )
+import Cardano.Metadata.Types.Common
+    ( File, fileContents )
+import Cardano.Metadata.Validation.Rules
+    ( Transform, ValidationError (ErrorCustom), defaultRules, mkTransform )
+import Cardano.Metadata.Validation.Types
+    ( Difference (Added, Changed, Removed), invalid, valid )
+import qualified Cardano.Metadata.Validation.Types as Common
+    ( Metadata )
+import Data.List.NonEmpty
+    ( NonEmpty )
+import Data.Text
+    ( Text )
+import Data.Validation
+    ( Validation )
+import Prelude hiding
+    ( log )
+
+import qualified Data.Aeson as Aeson
+import qualified Data.Aeson.Types as Aeson
+import qualified Data.Text as T
+
+data WalletValidationError
+  = WalletFailedToParseRegistryEntry Aeson.Value String
+  -- ^ Failed to parse a metadata entry from the file contents (json value, err)
+  | WalletFailedToValidate PartialGoguenRegistryEntry Text
+  -- ^ Failed to validate the metadata entry (json value, err)
+
+-- | Pretty print a wallet-specific validation error.
+prettyPrintWalletValidationError :: WalletValidationError -> Text
+prettyPrintWalletValidationError (WalletFailedToParseRegistryEntry val err) =
+  "Failed to parse wallet metadata entry from JSON: '" <> T.pack (show val) <> "', error was: " <> T.pack err <> "."
+prettyPrintWalletValidationError (WalletFailedToValidate val err) =
+  "Failed to validate wallet metadata entry '" <> T.pack (show val) <> "', error was: '" <> err <> "'."
+
+-- | The wallet validation rules consist of the default validation
+-- rules for metadata, as well as some wallet-specific validation (see
+-- @walletValidation@).
+walletValidationRules
+  :: Transform
+       (Difference (File Common.Metadata))
+       (Validation (NonEmpty (ValidationError WalletValidationError)))
+       ()
+walletValidationRules =
+  defaultRules
+  *> mkTransform walletValidation
+
+-- | Try to parse and validate a wallet metadata entry (see
+-- @parseRegistryEntry@ and @validateEntry@).
+walletValidation
+  :: Difference (File Common.Metadata)
+  -> Validation (NonEmpty (ValidationError WalletValidationError)) ()
+walletValidation diff = do
+  -- Find the new file we are going to validate
+  let
+    mMeta =
+      case diff of
+        Added newFile     -> Just (fileContents newFile)
+        Removed _         -> Nothing
```

Comment:
I thought removing from the registry was forbidden ?

Reply:
The composition of rules should satisfy our requirements, each
individual rule need only satisfy part of our requirements.

Removing from the registry is forbidden, but that's not the concern of
this rule. For example, the "metadata-lib" project provides a
"isJSONFile" rule here
(https://github.com/input-output-hk/metadata-server/blob/aebd96c623aaf32426e9b7189d60728fd3d6b768/metadata-lib/src/Cardano/Metadata/Validation/Rules.hs#L111).
That rule is only concerned with checking that each file has the
".json" extension. From the rule's point of view, it's worth
considering the question, "does this rule care?". In the case of
"isJSONFile", it wants to check that new files have the ".json"
extension, not check that you're only deleting ".json" files, that's a
very different rule. So it consciously decides to allow removed files,
knowing that it's not it's job. 

By keeping each rule "singular", composition of rules is frictionless.
Testing is also simplified.

In the proposed system, the GitHub validator checks that there is no
removal of files. There are already murmurs of allowing removal of
files under some trust scheme, so I think it best to keep that removal
logic in one place.